src/sqlfluff/dialects/dialect_mysql.py

```diff
@@ -249,6 +250,20 @@
     SingleIdentifierGrammar=ansi_dialect.get_grammar("SingleIdentifierGrammar").copy(
         insert=[Ref("SessionVariableNameSegment")]
     ),
+    ArithmeticBinaryOperatorGrammar=OneOf(
+        Ref("PlusSegment"),
+        Ref("MinusSegment"),
+        Ref("DivideSegment"),
+        Ref("MultiplySegment"),
+        Ref("ModuloSegment"),
+        Ref("BitwiseAndSegment"),
+        Ref("BitwiseOrSegment"),
+        Ref("BitwiseXorSegment"),
+        Ref("LogicalXorSegment"),
+        Ref("BitwiseLShiftSegment"),
+        Ref("BitwiseRShiftSegment"),
+        Ref("AmpersandAmpersandSegment"),
```

Comment:
I wonder do these two new ones belong in `ArithmeticBinaryOperatorGrammar` or `BooleanBinaryOperatorGrammar`? Not sure there's a massive difference between the two to be honest but `AND` and `OR` are in `BooleanBinaryOperatorGrammar`.

In fact, to make it even simpler, could you just override these:

```py
    AndKeywordSegment=StringParser("and", KeywordSegment, type="binary_operator"),
    OrKeywordSegment=StringParser("or", KeywordSegment, type="binary_operator"),
```

To these?

```py
    AndKeywordSegment=OneOf(
        StringParser("AND", KeywordSegment, type="binary_operator"),
        StringParser("&&", CodeSegment, name="double_ampersand" type="binary_operator"),
    ),
    OrKeywordSegment=OneOf(
        StringParser("OR", KeywordSegment, type="binary_operator"),
        StringParser("||", CodeSegment, name="double_pipe" type="binary_operator"),
        StringParser("XOR", KeywordSegment, type="binary_operator"),
    ),
```

(note I've also added `||` that you didn't include).

In fact if you could figure out where you want to insert `!`, you could do that similarly by replacing those `"NOT"`s with a similar `NotKeywordSegment`, set it to `NOT` in ANSI, and to `OneOf(...NOT...!....)` in MySQL.